File: Week 2/cpp/3.find_character_case.cpp
/*
 * WEEK 2 - C++ CONTROL FLOW
 * Topic: Character Case Detection via ASCII Values
 * File: 3.find_character_case.cpp
 *
 * PROBLEM:
 *  Output 1 for uppercase, 0 for lowercase, -1 otherwise.
 *
 * CONCEPT:
 *  Use the ASCII value of the character. Cast char to int (implicit) and
 *  test against the appropriate ranges. <cctype> offers `std::isupper`,
 *  `std::islower`, `std::isalpha` helpers.
 *
 * KEY POINTS:
 *  - 'A'-'Z' have codes 65-90; 'a'-'z' have 97-122.
 *  - `std::isupper((unsigned char) ch)` is the safe portable test.
 *  - Always cast to `unsigned char` before <cctype> functions to avoid UB
 *    when char is signed and the value is negative.
 *
 * SYNTAX:
 *   if (std::isupper((unsigned char) ch)) ...
 *
 * DRY RUN:
 *  ch='Q' -> isupper -> 1
 *  ch='q' -> islower -> 0
 *  ch='7' -> neither -> -1
 *
 * COMPLEXITY: O(1).
 */

#include <iostream>
#include <cctype>
#include <string>

int classify(char ch) {
    auto uc = static_cast<unsigned char>(ch);
    if (std::isupper(uc)) return 1;
    if (std::islower(uc)) return 0;
    return -1;
}

int main() {
    std::string s;
    if (!(std::cin >> s)) {
        // No input piped -- run the small demo.
        for (char c : {'Q', 'q', '7', '!'}) {
            std::cout << "'" << c << "' -> " << classify(c) << '\n';
        }
        return 0;
    }
    std::cout << classify(s[0]) << '\n';
    return 0;
}

/*
 * NOTES:
 *  - Java offers Character.isUpperCase / isLowerCase; C++ uses <cctype>.
 *  - char in C++ may be signed OR unsigned (implementation-defined).
 *    Casting to `unsigned char` first is the standard guard.
 *  - For locale-aware classification use <locale> or <cwctype> for wide chars.
 */

File: Week 2/python/3.find_character_case.py
"""
WEEK 2 - PYTHON CONTROL FLOW
Topic: Character Case Detection via ASCII Values
File: 3.find_character_case.py

PROBLEM:
Given a single character print:
  1  -> if the character is an UPPERCASE letter (A-Z)
  0  -> if the character is a LOWERCASE letter (a-z)
 -1  -> otherwise

CONCEPT:
Use the ASCII (Unicode) code point of the character. In Python `ord(ch)`
returns the int code point of a length-1 string. ASCII ranges:
  uppercase 'A'-'Z' -> 65-90
  lowercase 'a'-'z' -> 97-122

KEY POINTS:
 - `ord(ch)` -> int code point.
 - Chained comparisons make range tests pleasant: `'A' <= ch <= 'Z'`.
 - You can also use `str.isupper()`, `str.islower()`, `str.isalpha()`.

SYNTAX:
 if 'A' <= ch <= 'Z': ...
 if ch.isupper(): ...

DRY RUN:
 ch = 'q' -> ord = 113 -> in 97..122 -> print 0
 ch = 'A' -> ord = 65  -> in 65..90  -> print 1
 ch = '7' -> ord = 55  -> neither    -> print -1

COMPLEXITY: O(1).
"""

import sys


def classify(ch: str) -> int:
    """Return 1 for uppercase, 0 for lowercase, -1 otherwise."""
    if not ch:
        return -1
    # Prefer the readable, idiomatic test:
    if ch.isupper():
        return 1
    if ch.islower():
        return 0
    return -1

    # Equivalent ASCII-range test (kept for clarity):
    # if 'A' <= ch <= 'Z': return 1
    # if 'a' <= ch <= 'z': return 0
    # return -1


def main() -> None:
    if sys.stdin.isatty():
        # Demo path with a few hard-coded cases when no input is piped
        for c in ("Q", "q", "7", "!"):
            print(f"{c!r} -> {classify(c)}")
        return

    raw = sys.stdin.read().strip()
    ch = raw[0] if raw else ""
    print(classify(ch))


if __name__ == "__main__":
    main()


# NOTES:
# - Java has `Character.isUpperCase(ch)` etc.; Python's str methods are the equivalent.
# - Python strings are Unicode -- isalpha() returns True for letters beyond ASCII too.
# - Chained comparison `'A' <= ch <= 'Z'` is unique to Python (and a few other langs).
# - `ord('A') == 65`, `chr(65) == 'A'` -- bidirectional helpers.
